bigquery/scheduled_queries/append-monthly-users-from-cloudfront-logs.sql

```diff
@@ -1,96 +1,155 @@
   #Recalculates the table of monthly users taken from Cloudfront logs for each monthly user within this month,
   #and appends that to monthly users from previous months (which won't change over the course of the month).
   #Should overwrite the CALCULATED_monthly_users_from_cloudfront_logs table daily to accomplish this.
-  (
+WITH
+  cloudfront_logs AS (
   SELECT
-    month,
-    type,
-    device_category,
-    search_parameters,
-    vacancies_viewed_slugs,
-    vacancies_with_gmi_clicks_ids,
-    utm_sources,
-    utm_campaigns,
-    utm_mediums,
-    job_alert_destination_links,
-    job_alert_destinations,
-    referrers,
-    vacancies_viewed_slugs IS NOT NULL AS viewed_a_vacancy,
-    vacancies_with_gmi_clicks_ids IS NOT NULL AS clicked_get_more_information,
-    "subscription" IN UNNEST(utm_sources) AS from_job_alert,
-    ARRAY_LENGTH(search_parameters) AS unique_searches,
-    ARRAY_LENGTH(vacancies_viewed_slugs) AS vacancies_viewed,
-    ARRAY_LENGTH(vacancies_with_gmi_clicks_ids) AS vacancies_with_gmi_clicks
+    DATE_TRUNC(date, MONTH) AS month,
+    DATETIME(date,
+      time) AS datetime,
+    c_ip,
+    cs_uri_stem,
+    cs_referer,
+    c_user_agent,
+    cs_uri_query,
+    sc_content_type
+  FROM
+    `teacher-vacancy-service.production_dataset.cloudfront_logs`
+  WHERE
+    #only append users from last month (if it's the first day of the month) or this month (for days 2-31)
+    DATE_TRUNC(date, MONTH) = DATE_TRUNC(CURRENT_DATE() - 1, MONTH) ) (
+  SELECT
+    *,
+    CASE
+      WHEN utm_campaign LIKE "%alert%" THEN "Job alert"
+      WHEN LOWER(utm_medium) LIKE "%email%" THEN "Email"
+      WHEN referrer LIKE "%facebook%" OR referrer LIKE "%twitter%" OR referrer LIKE "%t.co%" OR referrer LIKE "%linkedin%" OR referrer LIKE "%youtube%" THEN "Social"
+      WHEN (referrer NOT LIKE "%teaching-jobs.service.gov.uk%"
+      AND referrer NOT LIKE "%teaching-vacancies.service.gov.uk%"
+      AND referrer NOT LIKE "%google%"
+      AND referrer NOT LIKE "%bing%"
+      AND referrer NOT LIKE "%yahoo%"
+      AND referrer NOT LIKE "%aol%"
+      AND referrer NOT LIKE "%ask.co%")
+    OR utm_medium="referral" THEN "Referral"
```

Comment:
Is this `OR` correct? Do we want to assume that if the referrer is not like any of the above, then the medium is referral?

Reply:
I think technically it *is* correct, due to a quirk of how the SQL 'not like' handles null values (NULL NOT LIKE '%aol%' returns FALSE, weirdly). But I don't think that's clear to the reader, so I'll add in an explicit NOT NULL check here as well.

Re. the logic itself, the data here is pretty messy, so we're trying to work out two possible ways traffic could be marked as referral traffic:
- the utm_medium parameter, which is where a referrer explicitly chooses to mark a link as a referral link
- the HTTP referrer

The HTTP referrer is often hidden, and most referrers won't add a utm_medium, hence why I'm using an OR here.